workload/schemachange: fix `CREATE SCHEMA` for the DSC

chrisseto opened the pull request:
### This PR is stacked on top of others!

Please consider reviewing those PRs first. This message will be removed once prerequisite PRs are merged and this one is rebased.

* 6e1ea5daac3f1f49b47660d3c7ea4e7b298b3ac4 is from #115357
* 230b58829c04940ede072ccf62ef35c9f5912011 is from #115358
* 4cc9b458821be299cae57c4feb1c0a37a8a52b00 is from #115359
* be063ef99c6c20b6661e918a455ee0b0315378f8 is from #115360
* 4e89e34921633d6918801e24ce895b2fe6b5887f is from #116788
* 95ffcf453bb30cc1a60b7a23428403e3a1a2091a is from #116789

---

#### 89e761126789d2828e141339c34e739203da2310 workload/schemachange: fix `CREATE SCHEMA` for the DSC

Previously, `CREATE SCHEMA` would break if executed with the declarative
schema changer due to the inclusion of `AUTHORIZATION`.

This commit removes the `AUTHORIZATION` clause from `CREATE SCHEMA`
operations until it's support within the DSC.

Epic: none
Release note: None

chrisseto committed:
workload/schemachange: implement `DROP FUNCTION`

This commit adds support for the `DROP FUNCTION` DDL to the RSW.

Epic: CRDB-19168
Informs: CRDB-3265

chrisseto committed:
workload/schemachange: implement `ALTER FUNCTION`

This commit adds support for the `ALTER FUNCTION RENAME TO` and the `ALTER
FUNCTION SET SCHEMA` DDLs to the RSW.

Other incantations of `ALTER FUNCTION`, namely `OWNER TO` and modifiers,
such as `STRICT`, have been omitted as they seem less likely to discover
schema related issues.

Epic: CRDB-19168
Informs: CRDB-3265
Release note: None

chrisseto committed:
workload/schemachange: implement `COMMENT ON`

This commit adds support for attaching comments to various schema
elements in the RSW. Notably, commenting on databases has been omitted
as the workload operates within the scope of a given database.

Unfortunately, `COMMENT ON` will consistently trigger the bug reported
in #72820. Therefore it is disabled by default.

Epic: CRDB-19168
Informs: CRDB-3265
Informs: #72820
Release note: None

chrisseto committed:
workload/schemachange: dump system tables to disk

Upon failure, the RSW would dump the namespace and descriptor tables to
the test log. While useful debugging information, it was rarely in a
format that was immediately useable to humans and it also obscured any
other log lines that would be emitted by the workload.

This commit changes the behavior to instead dump the tables to disk and
additionally to dump a JSONified version of the descriptors table which
can be audited manually.

Epic: none
Release note: None

chrisseto committed:
workload/schemachange: fix `CREATE SCHEMA` for the DSC

Previously, `CREATE SCHEMA` would break if executed with the declarative
schema changer due to the inclusion of `AUTHORIZATION`.

This commit removes the `AUTHORIZATION` clause from `CREATE SCHEMA`
operations until it's support within the DSC.

Epic: none
Release note: None

chrisseto committed:
workload/schemachange: deflake `ALTER PRIMARY KEY`

Previously, the `ALTER PRIMARY KEY` DDL suffered from a handful of
flakes and bugs. This commit is a collection of fixes thereof.

* Fix a swapped boolean in `UnderGoingSchemaChanges` queries.
* Correct error code for the `NonExistentColumn` case.
* Exclude columns with non-indexable data types.
* Don't rely on table statistics for uniqueness checks.
* Exclude generated columns.

The factor was a bit heavier than expected but the resultant code should
be easier to maintain and extend as need be.

Epic: none
Relese note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/116790)
<!-- Reviewable:end -->

chrisseto committed:
workload/schemachange: disable recently added DDL

This commit disables a handful of recently added operations in the RSW.
Despite numerous test and stress runs there are still plenty of flakes
floating around. Known flakes have been documented as issues and linked
to from `optype.go`

Unfortunately, #116799 may not correlate to a specific DDL and could
still pop up during stress runs. This commit includes some speculative
`WithStack` wrappings to help track down the cause.

Epic: none
Release note: None

rafiss reviewed (approved):
lgtm! this should help with some of the flakes

bors r+
